### DISCUSSION TITLE
Consider supporting regular expressions

### DISCUSSION THREAD

**Discussion by springcomp** [jmespath-community/jmespath.spec#23] [Ideas]
There seems to be some demand for supporting regular expressions.

- https://github.com/jmespath/go-jmespath/issues/64
- https://github.com/jmespath/jmespath.py/issues/213

This presents some challenges as there are many different _dialects_ of regexp, although there is an effort around
specifying an [interoperable regex](https://github.com/ietf-wg-jsonpath/iregexp) format.

Also, regular expressions uses cases fall into three main categories:
- Matching patterns, which could be useful in `filter-expressions`. It either matches, or it does not.
- Replacing substrings.
- Capturing and extracting substrings in named (or numbered) groups.

So a JEP would need to come up with intuitive syntax for those use cases that would be relevant.

The interoperable regex format proposal referred to above only supports the first scenario. So maybe that’s the only scenario that requires extending the JMESPath syntax.

The last two scenarios may be supported using dedicated functions.

**Comment by springcomp** [comment]
Tu use regular expressions in a `filter-expression` we would need:

- A new function `regex_match` defined like so:

```
boolean regex_match(string $text, regexp $regex)
```

`regexp` could be a string or could be a new datatype that could only be used in functions.

```abnf
regexp          = …
function-arg =/ regexp
```

## Example

Consider the following JSON document

```json
{
	"assets": [
		{ "name": "first-name" },
		{ "name": "last-name" },
		{ "name": "other" }
	]
}
```

To select all `assets` whose name matches a given pattern, the following hypothetical JMESPath expression could be used:

```
assets[? regex_match(name, /\-name$/i) ]
```

This would evaluate to:

```
[ { "name": "first-name"}, {"name": "last-name"} ]
```

**Comment by springcomp** [comment]
To support regular expressions to replace substrings, we would need a new function:

```
string regex_replace(string $text, regexp $regex)
```

For instance, using the same example as the previous comment, the following expression:

```
assets.{raw: regex_replace(name, /\-name$/g, '')}
```


would evaluate to:

```
[
  { "raw": "first" },
  { "raw": "last" },
  { "raw": "other" }
]
```

**Comment by springcomp** [comment]
Finally, a function could return the named or numbered captures from a regex.

```
object regex_capture(string $text, regexp $regex)
```

Example:

The simplest usage could look like:

```
regex_capure('hello, world!', /[a-z]+/)
```

Which evaluates to a JSON object if the match is successful:

```json
{ "0": "hello" }
```

Adding capturing groups would be supported like so:

```
regex_capure('hello, world!', /([a-z]+).*/)
```

Would return:

```json
{
  "0": "hello, world!",
  "1": "hello"
}
```
Likewise, this expression:
```
regex_capture('hello, world!', /(?<first>[a-z]+).*?(?<second>[a-z]+).*$/)
```

Would evaluate to:

```json
{
   "0": "hello, world!",
   "first": "hello",
   "second": "world"
}
```

This is currently not very useful, but in light of the probably upcoming function let, we could have expressions like so:

```
let( { re: regex_capture(foo, /^([^\-]+)?[a-z]+$/) }, &\<expression-type> )
```

**Comment by innovate-invent** [comment]
I like the idea of regular expression support, but think that it could be quite burdensome for some implementations. How would we handle varying RE specs (POSIX vs PCRE vs RE2) that are available for each language?

I think we should have an explicitly optional spec for the function interface to improve consistency between implementations, but leave it up to the implementers as to whether they want to implement it and what RE spec it would use.

**Reply by springcomp** [reply]
The more I think about this, the less I think we need a `regexp` type in the grammar. I think simple strings will be sufficient.

The reason why there is a `regexp` type in JavaScript is because it is an object and can be compiled. This is useful for `regexp` literals in the JavaScript source code. In JMESPath, we do not have a compilation step as the AST is interpreted in most implementations.

The following functions will be sufficient:

```
boolean regex_match(string $text, string $regex)
string regex_replace(string $text, string $regex)
object regex_capture(string $text, string $regex)
```

Where the `$regex` argument would be either implementation-specific, or using an interoperable subset of the regular expression language.

**Reply by innovate-invent** [reply]
I think for the spec we should require that error messages state that the function is either not implemented if that implementation does not support RE, or for any RE parsing error it must state the expected RE standard provided by that implementation. This makes it so that end users don't have to dig up what language their application is written in when trying to figure out if RE is supported and for what RE standard.

**Reply by springcomp** [reply]
I agree we may leave the door open to implementation-specific Regex dialects.

However, I think that, if specified, those functions should become mandatory for a standards compliant implementation to include those. I mean, an implementation can choose not to implement those functions, but they would not claim to be standards compliant.

**Reply by innovate-invent** [reply]
I think that would condemn all microarchitecture implementations to be painted with the broad "not compliant" brush. Some environments simply can't provide RE support.

**Comment by suntong** [comment]
Reiterating my view

> different programming languages have different regular expression implementations

Good point, but how about providing a regular expressions function among the [Built-in Functions](https://jmespath.org/specification.html#built-in-functions). 

The regular expressions _**will be Go specific**_, but that's much better than not having any regular expression support, or supporting them all, for `jp` at least.

the regular expression matching part has been fully implemented:
https://github.com/go-jsonfile/go-jmespath/commit/d90b399edd894622aaef0e529c0c54004aa02a9c#

just its exposed name need to be changed.

**Reply by springcomp** [reply]
Agreed. We are sort of [coming up to the same conclusion](https://github.com/jmespath-community/jmespath.spec/discussions/23#discussioncomment-3331117).

**Comment by Thaina** [comment]
I felt this should be fall into custom function/filter instead

Also there was also regex replace with custom function that would be great to supported

**Reply by springcomp** [reply]
Agreed, please see the [`regex_replace()`](https://github.com/jmespath-community/jmespath.spec/discussions/23#discussioncomment-2290775) function specified above.

**Comment by springcomp** [comment]
FYI: https://www.rfc-editor.org/rfc/rfc9485

**Reply by gibson042** [reply]
Thanks for this followup! Now that I-Regexp has been published as a stable document, I **fully** support adopting it here. There may still be demand for other dialects (including a possible future replacement of I-Regexp itself), so the design should plan for that possibility, but I see no reason not to move forward.

Do note the [details](https://www.rfc-editor.org/rfc/rfc9485#name-mapping-i-regexp-to-regexp-), though:
* matching is against "characters" that are Unicode scalar values (i.e., Unicode code points excluding surrogates U+D800 through U+DFFF)
* always anchored at both ends (e.g., string "aaa" is matched by pattern `.*a.*` but not by pattern `a`)
* absence of `\d`/`\s`/`\w` shorthand escapes and their `\D`/`\S`/`\W` complements
* absence of `\xHH`/`\uHHHH`/etc. escapes
* requirement of `\p{$category}` and `\P{$category}` escapes, but only for abbreviated General_Category values (e.g., not references to blocks/character properties/etc. and `\p{L}` and `\p{Lu}` are valid while `\p{Letter}` and `\p{Uppercase_Letter}` are not)
* `[^]` is not a valid character class
* `.` matches any character except U+000A LINE FEED ("\n") and U+000D CARRIAGE RETURN ("\r")
